src/topology/algebra/ordered.lean

```diff
@@ -1295,12 +1311,93 @@ begin
 
+/-- A bounded connected subset of a conditionally complete linear order includes the open interval
+`(Inf s, Sup s)`. -/
+lemma is_connected.Ioo_cInf_cSup_subset {s : set α} (hs : is_connected s) (hb : bdd_below s)
+  (ha : bdd_above s) :
+  Ioo (Inf s) (Sup s) ⊆ s :=
+λ x hx, let ⟨y, ys, hy⟩ := (is_glb_lt_iff (is_glb_cInf hs.nonempty hb)).1 hx.1 in
+let ⟨z, zs, hz⟩ := (lt_is_lub_iff (is_lub_cSup hs.nonempty ha)).1 hx.2 in
+hs.is_preconnected.Icc_subset ys zs ⟨le_of_lt hy, le_of_lt hz⟩
+
+lemma is_preconnected.Ioi_cInf_subset {s : set α} (hs : is_preconnected s) (hb : bdd_below s)
+  (ha : ¬bdd_above s) :
+  Ioi (Inf s) ⊆ s :=
+begin
+  have sne : s.nonempty := @nonempty_of_not_bdd_above α _ s ⟨Inf ∅⟩ ha,
+  intros x hx,
+  obtain ⟨y, ys, hy⟩ : ∃ y ∈ s, y < x := (is_glb_lt_iff (is_glb_cInf sne hb)).1 hx,
+  obtain ⟨z, zs, hz⟩ : ∃ z ∈ s, x < z := not_bdd_above_iff.1 ha x,
+  exact hs.Icc_subset ys zs ⟨le_of_lt hy, le_of_lt hz⟩
+end
+
+lemma is_preconnected.Iio_cSup_subset {s : set α} (hs : is_preconnected s) (hb : ¬bdd_below s)
+  (ha : bdd_above s) :
+  Iio (Sup s) ⊆ s :=
+@is_preconnected.Ioi_cInf_subset (order_dual α) _ _ _ s hs ha hb
+
+/-- A preconnected set in a conditionally complete linear order is either one of the intervals
+`[Inf s, Sup s]`, `[Inf s, Sup s)`, `(Inf s, Sup s]`, `(Inf s, Sup s)`, `[Inf s, +∞)`,
+`(Inf s, +∞)`, `(-∞, Sup s]`, `(-∞, Sup s)`, `(-∞, +∞)`, or `∅`. The converse statement requires
+`α` to be densely ordererd. -/
+lemma is_preconnected.mem_intervals {s : set α} (hs : is_preconnected s) :
+  s ∈ ({Icc (Inf s) (Sup s), Ico (Inf s) (Sup s), Ioc (Inf s) (Sup s), Ioo (Inf s) (Sup s),
+    Ici (Inf s), Ioi (Inf s), Iic (Sup s), Iio (Sup s), univ, ∅} : set (set α)) :=
+begin
+  rcases s.eq_empty_or_nonempty with rfl|hne,
+  { apply_rules [or.inr, mem_singleton] },
+  have hs' : is_connected s := ⟨hne, hs⟩,
+  by_cases hb : bdd_below s; by_cases ha : bdd_above s,
+  { rcases mem_Icc_Ico_Ioc_Ioo_of_subset_of_subset (hs'.Ioo_cInf_cSup_subset hb ha)
+      (λ x hx, ⟨cInf_le hb hx, le_cSup ha hx⟩) with hs|hs|hs|hs,
+    { exact (or.inl hs) },
+    { exact (or.inr $ or.inl hs) },
+    { exact (or.inr $ or.inr $ or.inl hs) },
+    { exact (or.inr $ or.inr $ or.inr $ or.inl hs) } },
+  { refine (or.inr $ or.inr $ or.inr $ or.inr _),
+    cases mem_Ici_Ioi_of_subset_of_subset (hs.Ioi_cInf_subset hb ha) (λ x hx, cInf_le hb hx)
+      with hs hs,
+    { exact or.inl hs },
+    { exact or.inr (or.inl hs) } },
+  { iterate 6 { apply or.inr },
+    cases mem_Iic_Iio_of_subset_of_subset (hs.Iio_cSup_subset hb ha) (λ x hx, le_cSup ha hx)
+      with hs hs,
+    { exact or.inl hs },
+    { exact or.inr (or.inl hs) } },
+  { iterate 8 { apply or.inr },
+    exact or.inl (hs.eq_univ_of_unbounded hb ha) }
+end
+
+/-- A preconnected set is either one of the intervals `Icc`, `Ico`, `Ioc`, `Ioo`, `Ici`, `Ioi`,
+`Iic`, `Iio`, or `univ`, or `∅`. The converse statement requires `α` to be densely ordererd. -/
```

Comment:
Maybe add a comment on the empty set, as it does not appear in the list below explicitly. Even if this is not necessary, I think I would include it, by the way, just for clarity. Same thing for `set_of_is_preconnected_eq_of_ordered`.